File: array-n-strings/is-palindrome.py
def is_palindrome(s):
    lo, hi = 0, len(s)-1

    while lo <= hi:
        if s[lo] == s[hi]:
            lo += 1
            hi -= 1
        else:
            return False

    return True


if __name__ == "__main__":

    values = {
        "palindrome": [
            "abccba",
            "aba",
            "012345543210"
        ],
        "non_palindrome": [
            "abc",
            "home"
        ]
    }

    for value in values["palindrome"]:
        assert is_palindrome(value)

    for value in values["non_palindrome"]:
        assert not is_palindrome(value)

File: array-n-strings/is_palindrome.rs
fn is_palindrome(s: &str) -> bool {
    let result: String = s
        .to_lowercase()
        .chars()
        .filter(|c| c.is_alphanumeric())
        .collect();

    result.chars().rev().collect::<String>() == result
}

fn main() {
    assert!(is_palindrome("A man, a plan, a canal, Panama!"));
    assert!(is_palindrome("Was it a car or a cat I saw?" ));
    assert!(is_palindrome("No 'x' in Nixon"));
    assert!(! is_palindrome("This is not palindrome"));
}
